VERSIONING.md

```diff
@@ -0,0 +1,108 @@
+  After `v1.2.1` is released, `v1.2` now points to `v1.2.1`.
+
+- `v<major>` always points to `v<major>.X` where X is the latest (largest) minor
+  version number. E.g. assume `v1` points `v1.2.0` initially. After `v1.3.0` is
+  released, `v1` now points to `v1.3.0`. After `v1.3.1` is released, `v1` now
+  points to `v1.3.1`.
+
+Note: We do NOT support the `latest` tag, since we cannot provide any
+compatibility guarantee for it, and the pipeline won’t produce deterministic
+results.
+
+## User-facing Surfaces
+
+All functions in the catalog comply with the [function spec]. To learn more
+about the functions concept,
+see [here](http://kpt.dev/book/02-concepts/03-functions).
+
+There are 2 user-facing surfaces for a function:
+
+- The `functionConfig`
+- The function itself
+
+### functionConfig Surface
+
+A `functionConfig` can be either a core resource (e.g. `ConfigMap`) or a custom
+resource. If the `functionConfig` is a CRD, it can be versioned independently as
+a normal CRD.
+
+### Function Surface
+
+#### What are NOT part of the function surface
+
+- The formatting of serialization for output items and results. e.g. yaml
+  indentation and order of fields in a map.
+- The order of resources in the output items.
+- The order of result items in the results.
+- The content of the unstructured messages in the results.
+
+#### What are part of the function surface
+
+- The supported fields in the ConfigMap as the `functionConfig`.
+- The supported versions of the CRD `functionConfigs`.
+- How the function behave given the input items and `functionConfig`:
+    - The reminder aspects of the output items that are not mentioned in the
+      previous section.
+    - The reminder aspects of the results that are not mentioned in the previous
+      section.
+
+### Breaking Changes
+
+We define a breaking change as: For any given input (including `input items`,
+`functionConfig` and OpenAPI), the function produces a different output
+(including output items, results) that are part of the user-facing surface.
+
+### Backwards Compatibility
+
+For versions after v1.0.0, we will:
+
+- Bump major version: There are breaking changes.
+
+- Bump minor version: There are backward-compatible features.
+
+- Bump patch version: There are only bug fixes and security fixes (e.g.
+  dependency package non-breaking version bump and base image non-breaking
+  version bump).
+
+For versions before v1.0.0, the major version is always `0` and we will:
+
+- Bumping minor version: There are breaking changes.
+
+- Bumping patch version: In all other cases, including backward-compatible
+  features, bug fixes and security fixes.
+
+Users won’t observe breaking changes if they are using the shorter semantic
+versions (e.g. `v1.2` and `v1`) and they can automatically get the latest secure
+patch version for free.
+
+## Best Practices
+
+- It is recommended to use shorter semantic versions (e.g. `v1.2` and `v1`) in
```

Comment:
I think we may want to talk about the benefits of each scheme:

Fully specified:
- Immutable infrastructure
- You are in control of when to update the functions

Floating
- Less maintaince
- Automatically pick up security and other fixes

Then let the reader decide for themselves.

Reply:
This is the most important bit for the function consumers, so listing what Frank listed above is a good starting point.

Reply:
To make this more explicit, I think we should remove "recommend to use shorter versions". It's a trade-offs between different types of risk:
- Risk of not having security patches picked up automatically
- Risk of being broken because we either accidentally pushed a breaking change or the consumer is dependent on a feature which is not part of the contract